### DISCUSSION TITLE
Ex 2: model files and forward signature

### DISCUSSION THREAD

**Discussion by waffelbaellchen** [sebastian-hofstaetter/teaching#70] [Advanced IR: Exercise Q&A] [answered]
Hi, in the provided model code files (e.g. `model_knrm.py`), the signature of the forward method states that `query `and `document` should be of type `Dict[str. torch.Tensor].` To my understanding, we expect query to be a dictionary with string-keys and torch.Tensor-values. However, in the forward method, the provided code accesses the values of `query` and `document` via
`query["tokens"]["tokens"] `
and
`document["tokens"]["tokens"]`

Maybe I misunderstand something, but shouldn't we access values via `query["tokens"]` or `document["tokens"]`? I do not quite understand how a torch.Tensor has a `["tokens"]` entry (which we try to access via `query["tokens"]["tokens"] `).

**Comment by sophiaalthammer** [comment, marked as answer]
Hi, I understand you confusion! The structure of query['tokens']['tokens'] is due how the IrTripleDatasetReader and the IrTupleDatasetReader (in data_loading.py) read in the textfields for the query_tokens, doc_pos_tokens and doc_neg_tokens. For each of the query, doc_pos, doc_neg it reads in a TextField (https://docs.allennlp.org/main/api/data/fields/text_field/) which is defined as TextFieldTensors = Dict[str, Dict[str, torch.Tensor]]. 

The given structure is correct and works with the predefined dataloaders, so in order to complete the missing parts, I would advise you to load a small example of data and look at the structure how it is read in.

Does that answer your question?

Best,
Sophia

**Reply by waffelbaellchen** [reply]
Yes, now it became clear, thank you!
